src/main/scala/ch/epfl/bluebrain/nexus/iam/index/InMemoryAclsTree.scala

```diff
@@ -0,0 +1,133 @@
+          case ((ownPaths, acc), entry @ (p, acl)) if containsOwn(acl)            => ownPaths + p -> (acc + entry)
+          case ((ownPaths, acc), (p, acl))                                        => ownPaths     -> (acc + (p -> acl.filter(identities)))
+        }
+      result
+    }
+
+    if (self) {
+      val result = if (ancestors) getWithAncestors(path) else get(path)
+      result.filter(identities).removeEmpty.pure
+    } else {
+      val result = removeNotOwn(getWithAncestors(path))
+      if (ancestors)
+        result.removeEmpty.pure
+      else
+        AccessControlLists(result.value.filterKeys(_.length == path.length)).removeEmpty.pure
+    }
+  }
+
+  private def getWithAncestors(path: Path): AccessControlLists = {
+    val currentAcls = get(path)
+    if (path.isEmpty) currentAcls
+    else currentAcls ++ getWithAncestors(path.tail)
+  }
+
+  private def pathOf(segments: Vector[String]): Path = Path(segments.mkString("/", "/", ""))
+
+  private def get(path: Path): AccessControlLists = {
+    val segments = path.segments.toVector
+
+    def inner(toConsume: Vector[String]): AccessControlLists = {
+      if (toConsume.contains(any)) {
+        val consumed = toConsume.takeWhile(_ != any)
+        tree.get(pathOf(consumed)) match {
+          case Some(children) if consumed.size + 1 == segments.size =>
+            AccessControlLists(acls.filterKeys(children.contains).toMap)
+          case Some(children) =>
+            children.foldLeft(AccessControlLists.empty) {
+              case (acc, (Segment(head, _))) =>
+                val toConsumeNew = (consumed :+ head) ++ segments.takeRight(segments.size - 1 - consumed.size)
+                acc ++ inner(toConsumeNew)
+              case (acc, _) => acc
+            }
+          case None => AccessControlLists.empty
+        }
+      } else {
+        val p = pathOf(toConsume)
+        acls.get(p).map(acl => AccessControlLists(p -> acl)).getOrElse(AccessControlLists.empty)
+      }
+    }
+    inner(segments)
+  }
+
+}
+
+object InMemoryAclsTree {
+
+  /**
+    * Constructs an in memory implementation of [[AclsIndex]]
+    *
+    * @tparam F the monadic type
+    */
+  final def apply[F[_]: Applicative](): InMemoryAclsTree[F] =
+    new InMemoryAclsTree[F](mutable.Map.empty[Path, Set[Path]], mutable.Map.empty[Path, AccessControlList])
+
+  implicit class MapMergeSyntax[K, V](private val map: mutable.Map[K, V]) extends AnyVal {
+
+    /**
+      * If the provided ''key'' is not already associated with a value, associates it with the given value.
+      * Otherwise, replaces the value with the results of the given ''f'' function. This operation is not atomic.
+      *
+      * @param key   the key with which the specified value is to be associated
+      * @param value the value to use if absent
+      * @param f     the function to recompute a value if present
+      */
+    def merge(key: K, value: V, f: V => V): Unit =
```

Comment:
This name is quite confusing, if `f` is a function which doesn't merge `value` and `c`.

Reply:
It is the same naming used from the [concurrentHashMap](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html) and they don't seem to feel is weird :)